### PR TITLE
Clipping if only one character text overflows

### LINKED ISSUE #99140: Text do not clip if only one character text overflows
### Code
```dart
import 'package:flutter/material.dart';

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();

  runApp(const MyHomePage(
    title: '',
  ));
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Scaffold(
        body: Center(
          child: SizedBox(
            width: 30,
            height: 200,
            child: ColoredBox(
              color: Colors.purple,
              child: Text(
                '7',
                textScaleFactor: 7,
                maxLines: 1,
                overflow: TextOverflow.clip,
              ),
            ),
          ),
        ),
      ),
    );
  }
}

```

### Expect
image

### Actual
image

### PR DESCRIPTION
**Pull request by xu-baolin**
Fixes #99140

### CODE REVIEW

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 1]
```diff
@@ -586,6 +586,14 @@ class TextPainter {
     return _paragraph!.didExceedMaxLines;
   }
 
+  /// The distance from the left edge of the leftmost glyph to the right edge of
+  /// the rightmost glyph in the paragraph.
+  ///
+  /// Valid only after [layout] has been called.
+  double get longestLine {
+    return _paragraph!.longestLine;
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/text_painter.dart:594]: add a `assert(!_debugNeedsLayout);`?

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 2]
```diff
@@ -644,7 +650,7 @@ class RenderParagraph extends RenderBox
     size = constraints.constrain(textSize);
 
     final bool didOverflowHeight = size.height < textSize.height || textDidExceedMaxLines;
-    final bool didOverflowWidth = size.width < textSize.width;
+    final bool didOverflowWidth = size.width < textSize.width || size.width < _textPainter.longestLine;
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:653]: Why would  `_textPainter.longestLine` ever be smaller than `textSize.width` (which is `_textPainter.size,width`) for a one char long text? That seems odd. Is there a bug in how textPainter calculates its actual width?

Comment (goderbauer) [reply packages/flutter/lib/src/rendering/paragraph.dart:653]: We also have `TextWidthBasis` which essentially changed `_textPainter.size.width` to be `_textPainter.longestLine`. I am not too familiar with that setting, though.

/cc @justinmc who added it a while back.

Comment (xu-baolin) [reply packages/flutter/lib/src/rendering/paragraph.dart:653]: > Why would `_textPainter.longestLine` ever be smaller than `textSize.width` (which is `_textPainter.size,width`) for a one char long text? That seems odd. Is there a bug in how textPainter calculates its actual width?

The `textPainter.size.width` always returns the layout `constraints.width`, this should be the implementation of the engine layer. @LongCatIsLooong  may know the details.

Comment (justinmc) [reply packages/flutter/lib/src/rendering/paragraph.dart:653]: _textPainter.longestLine is the width of the longest line of actual text.  _textPainter.width is either the same thing or the constraints width depending on TextWidthBasis ([code](https://github.com/flutter/flutter/blob/b7a18be0a027805705689827376c578d98d53012/packages/flutter/lib/src/painting/text_painter.dart#L538)) as you guys noted.

So it's possible and even common for _textPainter.longestLine to be shorter than _textPainter.width, for example:

Screen Shot 2022-03-22 at 9 23 40 AM

Here _textPainter.longestLine is just the width of that single character (a few pixels), while _textPainter.width is the width of the green Container (100px).

So the condition in the code here (`size.width < _textPainter.longestLine`) makes sense to me.  That will be true when the available space is less wide than a single character.

Comment (justinmc) [reply packages/flutter/lib/src/rendering/paragraph.dart:653]: PS Sorry it took me so long to respond here!  I'm bad with my notifications recently.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Just a nit about an assert.  The discussion around the width conditional seems ok to me.

**Comment by justinmc** [comment]
The failed Google tests were a few visual diff tests where the rightmost pixel of some text had a very slight, unnoticeable change in opacity.  It's slightly strange that this happened for text that wasn't a single character long, but probably just due to floating point math for the antialiasing.  I've set the Google tests to "passing", but we'll have to accept the new reference images after merge.

**Comment by xu-baolin** [comment]
> The failed Google tests were a few visual diff tests where the rightmost pixel of some text had a very slight, unnoticeable change in opacity. It's slightly strange that this happened for text that wasn't a single character long, but probably just due to floating point math for the antialiasing. I've set the Google tests to "passing", but we'll have to accept the new reference images after merge.

Interestingly, if this patch breaks things, it must be that `size.width < _textPainter.longestLine` is satisfied, and no clipping has occurred before. After clipping, the edges of the characters change slightly. I think this kind of scene is intended, right?

**Comment by justinmc** [comment]
I think it might be this: In the case where TextWidthBasis.longestLine is used and the parent's width tightly wraps the text, then `size.width` and `_textPainter.longestLine` should be equal.  However, due to some floating point error in their calculation, one is very very slightly different, and `size.width < _textPainter.longestLine` is true.  The text gets clipped, and some tiny subpixel change in the rendered text occurs.

Maybe we should try using something like [`applyFloatingPointHack`](https://github.com/flutter/flutter/blob/9da7317dac81630bd65672119a6e7e277c72c511/packages/flutter/lib/src/painting/text_painter.dart#L511) here and see if it fixes the Google tests?  Or maybe we should just consider it correct and say it should be clipped...

**Comment by xu-baolin** [comment]
> I think it might be this: In the case where TextWidthBasis.longestLine is used and the parent's width tightly wraps the text, then `size.width` and `_textPainter.longestLine` should be equal. However, due to some floating point error in their calculation, one is very very slightly different, and `size.width < _textPainter.longestLine` is true. The text gets clipped, and some tiny subpixel change in the rendered text occurs.

Something's not quite right here, in this scenario, `textSize.width` is no smaller than `_textPainter.longestLine` because of `applyFloatingPointHack`, it is not possible that `size.width >= textSize.width` but `size.width < _textPainter.longestLine`

Could you help to file a test to repro the google test? Or just make this breaking intended?

I have no idea make which double value hack to check the test now, or do you have any thoughts?

**Comment by xu-baolin** [comment]
@justinmc Hey, Could we merge this and accept the new golden file? : )

**Pull request merged**